append and remove from path

dhoehna opened the pull request:
This change adds two major API's
1. AppendToPath
2. RemoveFromPath

AppendToPath allows users to append to the Path Environment Variable.  The whole PATH variable will be tracked.

RemoveFromPath allows users to delete any path part they want to from the path.  The API checks to make sure that what the user wants to delete is in the PATH.  Additionally, the API makes sure that the path part the user wants to remove is a full item in the PATH and not a fragment.


There are other changes in this PR.  I'll explain them here.
 * General readability and code organization changes.
 * Removal of tests to run in different Trust Levels and Runtime Behaviors.  I will add them back after this PR.  Getting the above API changes to work AND trying to get the tests to run with different TL's and RB's caused too much confusion in the tests.

riverar commented on dev/Microsoft.Process.Environment/EnvironmentManager.cpp:
```diff
@@ -141,12 +136,147 @@ namespace winrt::Microsoft::ProjectReunion::implementation
+        std::wstring currentPath = GetPath();
+        std::wstring pathPartToFind(path);
+
+        if (pathPartToFind.back() != L';')
```
This throws if `RemoveFromPath` is called with an empty string.

dhoehna replied:
Nice catch.  Fixed.

DrusTheAxe commented on test/EnvironmentManagerTests/EnvironmentManagerUWPTests.h:
```diff
@@ -8,40 +8,47 @@ namespace ProjectReunionEnvironmentManagerTests
-        TEST_CLASS_CLEANUP(UWPRemoveEvs)
+        TEST_CLASS_CLEANUP(IWPRemoveEVs)
         {
-            RemoveProcessEV();
```
EV? Ah! EnvironmentVariable.

Minor Suggestion: Rename *RemoveEV(s) to *Cleanup() would avoid the abbreviation

dhoehna commented on dev/ChangeTracker/EnvironmentVariableChangeTracker.h:
```diff
@@ -41,52 +41,30 @@ namespace winrt::Microsoft::ProjectReunion::implementation
-            subKeyStream << "\\";
-            subKeyStream << m_Key;
-            subKeyStream << "\\";
+            HKEY topLevelKey;
```
Does Project Reunion have a case for, or against these kind of if statements?

DrusTheAxe replied:
Ternary operators are fine (no restrictive guidance like "never use if without {}")

Simple ternary operations like here simplify variable initialization e.g.
```
int foo{};
if (x)
{
    foo = y;
}
else
{
    foo = z;
}
```

vs

```
int foo{ x ? y : z };
```

This also enables `auto foo{ x ? y : z };` when auto makes sense. The former precludes auto since `auto foo{};` isn't viable

dhoehna commented on dev/ChangeTracker/EnvironmentVariableChangeTracker.h:
```diff
@@ -41,52 +41,30 @@ namespace winrt::Microsoft::ProjectReunion::implementation
-                    , keyToTrackChanges.put()
-                    , nullptr)));
-            }
+            THROW_IF_FAILED(HRESULT_FROM_WIN32(RegCreateKeyEx(HKEY_CURRENT_USER
```
That works.  I forgot there was a middle ground.

dhoehna replied:
I'm going through the rest of the code and changing my parameter list to this style.

riverar commented on dev/Microsoft.Process.Environment/EnvironmentManager.cpp:
```diff
@@ -141,12 +136,157 @@ namespace winrt::Microsoft::ProjectReunion::implementation
+            {
+                std::wstring pathPart(currentPath, left, right - left);
+
+                if (_wcsnicmp(pathPart.c_str(), pathPartToFind.c_str(), pathPart.size() + 1) == 0)
```
@DrusTheAxe  I think you're confusing the infamous l* comparison functions with wcscmp here, which performs an ordinal comparison. https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strcmp-wcscmp-mbscmp

DrusTheAxe replied:
Oddly enough I see email about a comment by @riverar but don't see the actual comment in the github view

>@DrusTheAxe I think you're confusing the infamous l* comparison functions with wcscmp here, which performs an ordinal comparison. https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strcmp-wcscmp-mbscmp

Yes the lstrcmp/etc functions are wrappers over CompareString etc lexical compares (https://docs.microsoft.com/windows/win32/intl/handling-sorting-in-your-applications) but I put those out of mind years ago :P

I was thinking of good old CRT wcs*(). Maybe I'm thinking of wcs* behavior varying dependent on locale? [][Sort Strings Ordinally](https://docs.microsoft.com/windows/win32/intl/handling-sorting-in-your-applications#sort-strings-ordinally) is pretty blunt about 'use CompareStringOrdinal' but vague on if wcs* are equivalent (and perhaps maybe, having changed over time?). Internally we've long had strong guidance "only use CompareStringOrdinal. Don't use wcs*" when handling package identity, application identity, filesystem, registry keys, and other ordinal-y data but perhaps things have changed?

@ShawnSteele are wcs* compare functions guaranteed to perform ordinal comparisons equivalent to `CompareStringOrdinal`?

riverar replied:
I quickly retracted it, because the behavior for the wide functions, when dealing with locale, gets really muddy and I think your original prescriptive guidance--replace the sucker!--is best. 😀

dhoehna replied:
This isn't the first time I've ran into this issue.  However, the epiphany just came to me.

Ordinal is binary.  Compares the bits.  This is when we want that sweet, sweet precisision.

Lexical, is based on the letters and numbers.  The out put, <0, 0, >0 is still the same, but what is compared is slightly different.

dhoehna commented on dev/Microsoft.Process.Environment/EnvironmentManager.cpp:
```diff
@@ -66,18 +66,13 @@ namespace winrt::Microsoft::ProjectReunion::implementation
-        {
-            THROW_HR(E_NOTIMPL);
-        }
+        throw hresult_not_implemented();
```
I thought a throw hresult_not_implemeneted was the standard sign of "This needs to be implemented and I haven't done it yet?"

EnvironmentManager will return empty defaults if IsSupported returns false.  This is because it relies on some platform features that are only is 21H1.

DrusTheAxe replied:
>I thought a throw hresult_not_implemeneted was the standard sign
Yes. Sorry, I thought you were closer to 'done' than perhaps you are

dhoehna replied:
Yeah.  I have a few more things left.
1. AppendToPathExecutable
2. RemoveFromPathExecutable
3. IsSupported.

dhoehna commented on dev/Microsoft.Process.Environment/EnvironmentManager.cpp:
```diff
@@ -120,15 +115,15 @@ namespace winrt::Microsoft::ProjectReunion::implementation
             {
                 THROW_IF_FAILED(HRESULT_FROM_WIN32(RegDeleteValue(environmentVariableKey.get()
                     , name.c_str())));
```
environmentVariableKey points to the registry location in either HKCU or HKLM for the environment variables.

I thought...it was a generally good idea to check in case a registry operation failed.

DrusTheAxe replied:
Generally one should have rigorous trapping of all error paths.

But deletion - removal - sometimes isn't blocking. Here if you fail to delete the regkey it's fatal. Is that the right answer?

At a minimum this fails if the regkey doesn't exist (ERROR_FILE_NOT_FOUND or ERROR_PATH_NOT_FOUND). That's wrong. See other comments on the subject

riverar replied:
Not sure I agree with the suggested change (to swallow the failure). The developer is calling this API to effect a change, a deletion in this case. While I agree, the end-goal is the same, the operation is not. A developer may want to handle the case where a registry value is unexpectedly missing--perhaps for auditing or some sort of misconfiguration UX. The suggested change would make handling that case impossible and knock it out of alignment with Win32 Registry APIs.

DrusTheAxe replied:
>Is that the right answer?

I didn't mean to suggest swallowing the error here. Just asking to think it through - is this required or best-effort?

But regardless the code needs changing, as ERROR_FILE_NOT_FOUND and ERROR_PATH_NOT_FOUND are error codes but not error conditions. You want the value gone, whether you delete it (ERROR_SUCCESS) or it's already gone (ERROR_FILE/PATH_NOT_FOUND) you get your wish.

riverar replied:
I don't understand. I have [beefbrain](https://hypnospace.fandom.com/wiki/Beefbrain). I think you are suggesting removal of THROW_IF_FAILED and effectively swallowing any non-success values. Am I missing something? Can you type out the exact change you're thinking about if that's not right?

DrusTheAxe replied:
No. At a minimum I'm stating the current code is bugged

```
THROW_IF_FAILED(HRESULT_FROM_WIN32(RegDeleteValue(environmentVariableKey.get(), name.c_str())));
```

and needs to be

```
const auto rc{ RegDeleteValue(environmentVariableKey.get(), name.c_str()) };
THROW_HR_IF(HRESULT_FROM_WIN32(rc), (rc != ERROR_SUCCESS) && (rc != ERROR_FILE_NOT_FOUND));
```

Right now the code throws if the reg value doesn't exist. But this should be treated as a success case, not a failure.

This is an common annoyance of the registry APIs, equally commonly solved by making small wrappers -- OpenIfExists, DeleteIfExists, etc to map the common not-found cases to ERROR_SUCCESS/S_OK so the WIL macros neatly handle it

riverar replied:
Got it, thanks. This is an _internal_ delete occurring as a result of a developer calling `EnvironmentManager::SetEnvironmentVariable` with a blank value. Agree on delete-if-exists semantics here.

(I wonder if it makes more sense for this API to throw/error on blank input, and explicitly add a `EnvironmentManager::RemoveEnvironmentVariable` method. Perhaps out of scope.)

dhoehna replied:
RemoveEnvironmentVariable is a neat idea.  I want to keep the API as close to a Win32 API as possible to make the transition seamless.  Both C++ and C# have a SetEnvironmentVariable feature and the way to remove a variable is by setting the input to empty.

dhoehna commented on dev/Microsoft.Process.Environment/EnvironmentManager.cpp:
```diff
@@ -120,15 +115,15 @@ namespace winrt::Microsoft::ProjectReunion::implementation
             {
                 THROW_IF_FAILED(HRESULT_FROM_WIN32(RegDeleteValue(environmentVariableKey.get()
```
Right.  Yes.  Thank you.

dhoehna commented on dev/Microsoft.Process.Environment/EnvironmentManager.cpp:
```diff
@@ -120,15 +115,15 @@ namespace winrt::Microsoft::ProjectReunion::implementation
-                reinterpret_cast<WPARAM>(nullptr), reinterpret_cast<LPARAM>(L"Environment"),
-                SMTO_NOTIMEOUTIFNOTHUNG | SMTO_BLOCK, 1000, nullptr);
+                LRESULT broadcastResult = SendMessageTimeout(HWND_BROADCAST, WM_SETTINGCHANGE,
+                    reinterpret_cast<WPARAM>(nullptr), reinterpret_cast<LPARAM>(L"Environment"),
```
It is a UINT pointer.

Why pull out the WPARAM?  To get rid of the cast?

dhoehna commented on dev/Microsoft.Process.Environment/EnvironmentManager.cpp:
```diff
@@ -120,15 +115,15 @@ namespace winrt::Microsoft::ProjectReunion::implementation
-                reinterpret_cast<WPARAM>(nullptr), reinterpret_cast<LPARAM>(L"Environment"),
-                SMTO_NOTIMEOUTIFNOTHUNG | SMTO_BLOCK, 1000, nullptr);
+                LRESULT broadcastResult = SendMessageTimeout(HWND_BROADCAST, WM_SETTINGCHANGE,
+                    reinterpret_cast<WPARAM>(nullptr), reinterpret_cast<LPARAM>(L"Environment"),
```
L"Enviornment"[0] is what I came up with.

dhoehna commented on dev/Microsoft.Process.Environment/EnvironmentManager.cpp:
```diff
@@ -120,15 +115,15 @@ namespace winrt::Microsoft::ProjectReunion::implementation
-            LRESULT broadcastResult = SendMessageTimeout(HWND_BROADCAST, WM_SETTINGCHANGE,
-                reinterpret_cast<WPARAM>(nullptr), reinterpret_cast<LPARAM>(L"Environment"),
-                SMTO_NOTIMEOUTIFNOTHUNG | SMTO_BLOCK, 1000, nullptr);
+                LRESULT broadcastResult = SendMessageTimeout(HWND_BROADCAST, WM_SETTINGCHANGE,
```
This does not work.  Wil complains that it is treating the LRESULT as a BOOL which is a no-go.

dhoehna commented on dev/Microsoft.Process.Environment/EnvironmentManager.cpp:
```diff
@@ -141,12 +136,157 @@ namespace winrt::Microsoft::ProjectReunion::implementation
+                if (result == 0)
+                {
+                    THROW_HR(HRESULT_FROM_WIN32(GetLastError()));
+                }
```
Done.  BTW.  It is THROW_IF_WIN32_BOOL_FALSE.

dhoehna commented on dev/Microsoft.Process.Environment/EnvironmentManager.cpp:
```diff
@@ -141,12 +136,157 @@ namespace winrt::Microsoft::ProjectReunion::implementation
+            if (m_Scope == Scope::Process)
+            {
+                BOOL result = FALSE;
+                result = ::SetEnvironmentVariable(L"Path", newPath.c_str());
```
True, true.   It is changed.

dhoehna commented on dev/Microsoft.Process.Environment/EnvironmentManager.cpp:
```diff
@@ -159,6 +299,46 @@ namespace winrt::Microsoft::ProjectReunion::implementation
+        {
+            wil::unique_hkey environmentVariableKey = GetRegHKeyForEVUserAndMachineScope();
+
+            DWORD sizeOfEnvironmentValue;
```
talking about the variable sizeOfEnvironmentValue?

That is true, most people use a suffix.  I prefer prefix because it matches what we would say in English.  When talking about code we say "This is the size of the environment value."  Of course, it could go the other way around and say "This is the environment value size."

I guess I think in the former than the latter.  However, this is kinda like hungarian notation where we know it is the size of something.

dhoehna commented on test/EnvironmentManagerTests/EnvironmentManagerCentennialTests.h:
```diff
@@ -8,42 +8,47 @@ namespace ProjectReunionEnvironmentManagerTests
         BEGIN_TEST_CLASS(EnvironmentManagerCentennialTests)
-            TEST_CLASS_PROPERTY(L"RunAs", L"{UAP,Restricted,Elevated,InteractiveUser}")
+            TEST_CLASS_PROPERTY(L"RunAs", L"UAP")
+            TEST_CLASS_PROPERTY(L"RunFixtureAs", L"ElevatedUser")
```
Yep.  I figured a way to get all the tests to run and get the fixture to run as well.

dhoehna commented:
Addressed comments.  One change the push 9fa3dc8 made was to have the tests run for different Trust levels.  Centennial has tests for both Partial and Full trust while Win32 runs for Elevated, Restricted, and LowIL.

The WIn32 Restricted tests are all blocked as TAEF does not have permission to make a test as restricted.  Besides that, all tests pass.
